rabbitmq/hooks/reconfigure

```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+exec 2>&1
+
+source "{{pkg.svc_config_path}}/env"
+
+{{#if cfg.rabbitmq.cluster.enabled ~}}
+# Make sure that RabbitMQ started before we attempt to cluster
+rabbitmqctl wait {{pkg.svc_var_path}}/db/rabbit@{{sys.hostname}}.pid
+{{#if svc.me.follower ~}}
```

Comment:
RabbitMQ has no concept of a leader or a follower. This should probably be changed to use the bind syntax rather than leader/follower.

Reply:
Do you have that syntax? I wasn't able to get anything related to `bind` working. It's been a few weeks and I don't have any notes of what I tried, though.

Reply:
```
{{#eachAlive bind.rabbitmq as |member| ~}}
rabbitmqctl join_cluster rabbit@{{member.sys.ip}}
{{/eachAlive ~}}
```

Something like this would attempt a `join_cluster` on every bound rabbitmq that is currently alive.

Reply:
If you just want to try the first one you can also say 
```
{{rabbitmqctl join_cluster rabbit@{{bind.rabbitmq.members.first.sys.hostname}}
```

Reply:
@elliott-davis I'd be careful with this syntax for now. `.first` will pick up confirmed dead and departed members. We plan to add the `firstAlive` field in the future, but it's not there yet :smile:

Reply:
While RabbitMQ doesn't have a leader/follower concept, I don't think that this is actually going to be the right approach, since it doesn't work.

I tried this with two startup scenarios. Both of them running on three docker containers and the package was built using the `eachAlive` block, like so, in the `reconfigure` hook.

```
{{#if cfg.rabbitmq.cluster.enabled ~}}
# Make sure that RabbitMQ started before we attempt to cluster
rabbitmqctl wait {{pkg.svc_var_path}}/db/rabbit@{{sys.hostname}}.pid
rabbitmqctl stop_app
{{#eachAlive bind.rabbitmq as |member| ~}}
rabbitmqctl join_cluster rabbit@{{member.sys.hostname}}
{{/eachAlive ~}}
rabbitmqctl start_app
```

### Scenario one: standalone topology

This requires that the configuration be present from the startup a la `ENV` variables, or that `hab config apply` is done against all members of the ring for the configuration to be applied. However, none of the members join the cluster in rabbitmq. In fact, the reconfigure hook doesn't have a `join_cluster` command written out from the `eachAlive` block in the template.

### Scenario two: leader/follower topology

This allows running `hab config apply` against only one supervisor, and it will apply the configuration to all members of the ring. However, they do not join the cluster - they also do not have the `join_cluster` command in the reconfigure hook.

### Conclusion

While it may not make sense to RabbitMQ, the reconfigure hook I wrote with the `if svc.me.follower`/`with svc.leader` block works to automatically have the cluster get created when `hab config apply` enables it with the configuration option.

If this is a blocker for merging this, someone else will have to spend time to figure it out, as I've lost track how much time I've spun my wheels trying to get the right syntax here.

Reply:
@jtimberman this is actually a very useful conversation for us - it helps us identify missing features and poor documentation. Answering why the proper syntax isn't obvious in such a small language is a really important part of this process. It's something that all @habitat-sh/habitat-core-plans-maintainers sign up for when they join the group!

After reviewing the plan a bit it looks like there was a bit of a misunderstanding or oversight when this plan was authored. Having an optional bind to yourself is not the best way to get at the members of your own group, we actually have that functionality baked in 😄. The syntax you're looking for is this:

```
{{#eachAlive svc.members as |member|}}
rabbitmqctl join_cluster rabbit@{{member.sys.hostname}}
{{/eachAlive ~}}
```

Using the boolean check to see if clustering is enabled is just fine before this stanza is hit is just fine. If you'd like to make the changes here and push this over the line it'd be great. If not, another @habitat-sh/habitat-core-plans-maintainers can tack the additional commits on this PR for you and see it over the finish line.

Reply:
@reset @elliott-davis @predominant When using the `#eachAlive` syntax on a reconfiguration event the habitat supervisors of all members in the ring are restarting rabbitmq at the same time. This means that when the `join_cluster` commands fire off from the `reconfigure` script, rabbitmq is not running, or the app is not fully operational yet.

When using `rabbitmqctl join_cluster`, it expects that rabbitmq is fully operational on the target system, and it doesn't retry on failure. We could attempt to do retry semantics in the reconfigure script, but the behavior we get out of habitat's leader/follower topology make it much easier to implement a working configuration. So while rabbitmq doesn't have the concept of "leader/follower" in its terminology, it does in its behavior.

Reply:
Oodles of back-scroll here and I'm just getting caught up. I'd say I'm pretty firmly in the camp of: If we're using a side-effect of a feature to get an expected behavior out of a plan then theres a high likelihood we're implementing an anti-pattern because of a lack of the correct feature. Which is to say, I don't ever want to blanket :-1: on a PR based on only principle, but reading through the backscroll I'm not entirely convinced that the leader/follower path actually solves the problem as much as it serves as a bandaid that could teach community members some less than desirable patterns in their plans.

All of this is to say I think @jtimberman you _might_ be on to a gap in featureset here. Perhaps something along the lines of daemonsets or the concept of a permanent peer thats specific to a servicegroup.

Reply:
@jtimberman in reading the rabbitmq cluster docs, you actually only need a single alive peer that is not yourself to cluster, and you need to be initialized first. I think leader/follower is actually the right pattern. They also have some logic around cluster behavior, in particular the "first" node in the cluster is often special. Something like:

1. start in leader/follower
2. All non-leader nodes join only the leader

There is no step 3.

Love,
Adam

Reply:
To be clear, I think what happens in @reset snippet above is that you are re-joining the same cluster a bunch of times.

Reply:
I spent some time last night running through this PR and an exercise on how to potentially do this without leader/follower. After hitting every pothole that @jtimberman probably hit I finally understand why this is a problem. Ultimately there is a race between the apps coming up on the Erlang VM so using leader follower allows the config change to work gracefully.

I think I got hung up on the semantics of leader follower and what RabbitMQ calls a distributed cluster. After going through RabbitMQ's magical forest I'm 👍  on merging this.

Reply:
Lets do it then.